src/osp/UserInputHandler.cpp

```diff
@@ -455,4 +466,113 @@ void UserInputHandler::scroll_delta(Vector2i offset)
+        std::vector<ButtonVarConfig> primary = parse_config_string(tab.second.at("primary").as_string().str);
+        controls.insert(controls.end(), primary.begin(), primary.end());
+
+        std::vector<ButtonVarConfig> secondary = parse_config_string((tab.second.at("secondary").as_string().str));
+        controls.insert(controls.end(), primary.begin(), primary.end());
+
+        //Will have to modify the holdable param so that it works
+        config_register_control(name, true,
+            controls);
+    }
+}
+
+std::vector<ButtonVarConfig> UserInputHandler::parse_config_string(std::string_view config)
+{
+    std::vector<ButtonVarConfig> conf;  
+    std::string string = std::string(config);
+    std::stringstream ss(string);
+    std::string token;
+    while (std::getline(ss, token, '+')) {
+        //Parse token
+        ButtonVarConfig button = get_button(token);
+        conf.push_back(button);
+    }
+
+    int size = conf.size() - 1;
+    conf[size].m_nextOp = ButtonVarConfig::VarOperator::OR;
+    conf[size].m_trigger = ButtonVarConfig::VarTrigger::PRESSED;
+
+    return conf;
+}
+
+ButtonVarConfig UserInputHandler::get_button(std::string_view text)
+{
+    using namespace osp;
+    using Key = OSPMagnum::KeyEvent::Key;
+    using Mouse = OSPMagnum::MouseEvent::Button;
+    using VarOp = ButtonVarConfig::VarOperator;
+    using VarTrig = ButtonVarConfig::VarTrigger;
+    ButtonVarConfig config(0, 0, ButtonVar::VarTrigger::HOLD, false, ButtonVar::VarOperator::AND);
+    std::cout << text << std::endl;
+    if (text == "RShift") {
+        config.m_devEnum = (int)Key::RightShift;
+    }
+    else if (text == "LShift") {
+        config.m_devEnum = (int)Key::LeftShift;
+    }
+    else if (text == "LCtrl") {
+        config.m_devEnum = (int)Key::LeftCtrl;
+    }
+    else if (text == "RCtrl") {
+        config.m_devEnum = (int)Key::RightCtrl;
+    }
+    else if (text == "LAlt") {
+        config.m_devEnum = (int)Key::LeftAlt;
+    }
+    else if (text == "RAlt") {
+        config.m_devEnum = (int)Key::RightAlt;
+    }
+    else if (text == "Up") {
+        config.m_devEnum = (int)Key::Up;
+    }
+    else if (text == "Down") {
+        config.m_devEnum = (int)Key::Down;
+    }
+    else if (text == "Left") {
+        config.m_devEnum = (int)Key::Left;
+    }
+    else if (text == "Right") {
+        config.m_devEnum = (int)Key::Right;
+    }
+    else if (text == "RMouse") {
+        config.m_devEnum = (int) Mouse::Right;
+        config.m_device = osp::sc_mouse;
+    }
+    else if (text == "LMouse") {
+        config.m_devEnum = (int)Mouse::Left;
+        config.m_device = osp::sc_mouse;
+    }
+    else if (text == "MMouse") {
+        config.m_devEnum = (int)Mouse::Middle;
+        config.m_device = osp::sc_mouse;
+    }
+    else if (text.length() == 1) {
+        //Ascii
+        int ascii = (int)text.at(0);
+        if (isalpha(ascii)) {
+            ascii = tolower(ascii);
+        }
+       config.m_devEnum = ascii;
```

Comment:
I hate to break it to you but this isn't going to work, the key enums can't be assumed to map to ASCII values. The `Magnum::Key` enums are mapped to the key enums from SDL2 and GLFW or any other OpenGL context manager that's implemented, so this part isn't safe. We'll have to figure out another way to map it.

Reply:
That's disappointing.
So the question now is, do I create a whole map for ascii values, because I'm afraid that that would look very ugly.

Reply:
It's usually reliable to use the enum:
```cpp
int keyEnum = (int)KeyEvent::Key::A + (tolower(text.at(0)) - 'a');
```
A map of ASCII values is not ugly, it's perfectly fine if it's nicely arranged and has a clear reason to be there. Since ASCII values and keycodes will never change, then a switch(char) is the best option here.

There are many other weird keys that don't have single-character names, and for that; a string to keycode&device map would be practical. See my other comment above on this.

Reply:
Good idea, that should work in most cases, even if it's not watertight. The main issue is that the keycodes map to keys and not letters, so GLFW chooses to call the keycodes by their lowercase ASCII value, and SDL2 chooses to call them by uppercase values. Indexing off of the enum value fixes it for those cases, at least.

Reply:
So this is a solution I came up with, which should take into account normal letters, except for special characters. We may need some sort of way to detect special characters such as &,*,@, etc.
```c++

std::tuple<int, int> settings::parse_key(std::string string)
{
    using namespace osp;
    using Key = OSPMagnum::KeyEvent::Key;
    using Mouse = OSPMagnum::MouseEvent::Button;
    using VarOp = ButtonVarConfig::VarOperator;
    using VarTrig = ButtonVarConfig::VarTrigger;
    
    //Device, key
    std::map<std::string, std::tuple<int, int>> m_strToButton = {
        {"RShift", {(int)osp::sc_keyboard, (int)Key::RightShift}},
        {"LShift", {(int)osp::sc_keyboard, (int)Key::RightShift}},
        {"LCtrl",{(int)osp::sc_keyboard, (int)Key::LeftCtrl}},
        {"RCtrl", {(int)osp::sc_keyboard, (int)Key::RightCtrl}},
        {"LAlt", {(int)osp::sc_keyboard, (int)Key::LeftAlt}},
        {"RAlt", {(int)osp::sc_keyboard, (int)Key::RightAlt}},
        {"Up", {(int)osp::sc_keyboard, (int)Key::Up}},
        {"Down", {(int)osp::sc_keyboard, (int)Key::Down}},
        {"Left", {(int)osp::sc_keyboard, (int)Key::Left}},
        {"Right", {(int)osp::sc_keyboard, (int)Key::Right}},
        {"RMouse", {(int)osp::sc_mouse, (int)Mouse::Right}},
        {"LMouse", {(int)osp::sc_mouse, (int)Mouse::Left}},
        {"MMouse", {(int)osp::sc_mouse, (int)Mouse::Middle}}
    };
    std::tuple<int, int> tup = {-1, -1};
    //Select device
    if (m_strToButton.count(string)) {
        //Contains, so get the button
        tup = m_strToButton[string];
    }
    else if(string.length() == 1){
        //Ascii
        if(isalpha(string.at(0)))
            tup = { (int)osp::sc_keyboard, ((int) Key::A + tolower(string.at(0)) - 'a') };
        else if (isdigit(string.at(0))) {
            tup = { (int)osp::sc_keyboard, ((int)Key::One + tolower(string.at(0)) - '1') };
        }
    }
    return tup;
}
```

Reply:
Honestly I think the best thing to do is just put all the keys in the map. Yeah it's kind of ugly, but it's a one-time thing, is guaranteed to work for all platforms that Magnum implements, and none of this code is performance sensitive or anything. 

In my opinion it's not worth the headache just to avoid 100 lines of keymap definition. Sometime interfaces just need a big list of mapping definitions and there's nothing you can do to avoid it, so it's not the end of the world, and I'm not sure if it's worth it to try and build something clever and programmatic if it's not guaranteed to work for every platform. Circumventing Magnum's abstraction layer is something to be avoided unless absolutely necessary.

Handling the special characters will require this treatment anyways so just putting everything in one big ugly map will simplify things substantially.

Reply:
There's a few issues here:

* A new map gets allocated every single time the function is called. This can be fixed my making the map static const. However, the [] operator does not work with const, and will modify maps if a value isn't found. Note the use of 'find' in other parts of the code, where it also searches the map only once.
* "m_" denotes a member variable. "sc_" would be the one for static const. There is a style guide that should have been put somewhere more accessible that we ~~very strictly~~ usually follow:
https://docs.google.com/document/d/1dJW-E6SN41XQuvidHRhp41O_1mJmP90bar1egZ31c10/edit?usp=sharing

I'd suggest adding some tab-aligned spaces to the map initialization to vertically align the columns into a more readable table.

Overall, I think the function is okay for the time being; it's already a couple hundred times faster than JS. It's best care more about the structure of the program vs. the insides of individual functions. Judging by the 'settings::' in the declaration, I'd assume loading functions have already been separated into this. My concern is that It's common in OOP to put everything an object can do in a single class, but here I'd highly recommended separating the loading functions from the loaded data. For example, Magnum has ImageData which can be created by multiple importers like StbImageImporter. ImageData is independent of StbImageImporter, and knows nothing of it.